llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,461 @@
+#include "Utils/AIELoopUtils.h"
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/CodeGen/LiveDebugVariables.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/LiveRegMatrix.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/MC/MCRegister.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <llvm/CodeGen/MachineBasicBlock.h>
+#include <map>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-wawreg-rewrite"
+
+namespace {
+
+cl::opt<bool> DisableCopySubRegHandling(
+    "aie-waw-disable-copy-sub-reg-handling",
+    cl::desc("Disable Copy Sub register handling, where WAW regs will not be "
+             "replaced, if reg parts are reused."),
+    cl::init(false), cl::Hidden);
+
+///
+/// This rewrites Registers, if they are redefined in the same Loop body and
+/// could cause WAW issues in the loop pipelining.
+/// The pass will update the \p VirtRegMap so that the new vregs have fixed
+/// assignments, guaranteeing that the reg renaming does not cause conflicts.
+class AIEWawRegRewriter : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIEWawRegRewriter() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<VirtRegMap>();
+    AU.addPreserved<VirtRegMap>();
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+};
+
+static bool
+renameMBBPhysRegs(const SmallVector<const MachineBasicBlock *, 4> &LoopBBs,
+                  const MachineFunction &MF, VirtRegMap &VRM,
+                  LiveRegMatrix &LRM, const MachineRegisterInfo &MRI,
+                  const LiveIntervals &LIS, const AIEBaseRegisterInfo &TRI);
+
+/// Find Free register of the same register Class type and
+// exclude the forbidden Physical register from the result
+static SmallVector<MCPhysReg, 16> getPriorityReplacementRegs(
+    const Register Reg, const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+    const MachineFunction &MF, const VirtRegMap &VRM, LiveRegMatrix &LRM,
+    const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+    const AIEBaseRegisterInfo &TRI);
+
+static bool replaceReg(const Register Reg,
```

Comment:
I think that most of those functions can be moved to the methods of the pass, and some parameters can be converted to class members. We are passing to much parameters to every function, and some of them are always the same. `const MachineRegisterInfo &MRI` for example...

Reply:
I moved all the Machine function derived objects and Analysispass results to the class attributes. The methods are now less crowded with arguments.